src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -55,51 +71,153 @@
 /**
  * Chooses manageable workspace names for branch projects.
  * @see "JENKINS-34564"
+ * @see "JENKINS-2111"
  */
 @Restricted(NoExternalUse.class)
 @Extension(ordinal = -100)
 public class WorkspaceLocatorImpl extends WorkspaceLocator {
 
     private static final Logger LOGGER = Logger.getLogger(WorkspaceLocatorImpl.class.getName());
 
+    @Deprecated
     static final int PATH_MAX_DEFAULT = 80;
-    /** The most characters to allow in a workspace directory name, relative to the root. Zero to disable altogether. */
-    // TODO 2.4+ use SystemProperties
+    /**
+     * The most characters to allow in a workspace directory name, relative to the root.
+     * Zero to disable altogether.
+     * Not used for new workspaces; only for continuing to use workspaces created prior to {@link #MAX_LENGTH}.
+     */
+    @Deprecated
     static /* not final */ int PATH_MAX = Integer.getInteger(WorkspaceLocatorImpl.class.getName() + ".PATH_MAX", PATH_MAX_DEFAULT);
 
+    static /* not final */ int MAX_LENGTH = Integer.getInteger(WorkspaceLocatorImpl.class.getName() + ".MAX_LENGTH", NameMangler.MAX_SAFE_LENGTH);
+
+    static /* not final */ boolean MULTIBRANCH_ONLY = Boolean.parseBoolean(System.getProperty(WorkspaceLocatorImpl.class.getName() + ".MULTIBRANCH_ONLY", /* TBD */ "true"));
```

Comment:
Could be set to `false` to turn on the smarter behavior for all projects, including standalone Pipeline and even freestyle, matrix, etc. On the one hand I do not like having installation of what is otherwise a library plugin affect behavior of core functionality. On the other hand, the name shortening is very useful for projects in deep folder hierarchies; the character sanitization is useful for all sorts of project generation systems other than multibranch; and the automatic cleanup is useful for pretty much everyone, since `WorkspaceCleanupThread` is unreliable.